spark/src/main/scala/io/substrait/spark/expression/ToWindowFunction.scala

```diff
@@ -0,0 +1,141 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.substrait.spark.expression
+
+import io.substrait.spark.expression.ToWindowFunction.{fromSpark, fromSparkFollowing, fromSparkPreceding}
+
+import org.apache.spark.sql.catalyst.expressions.{CurrentRow, Expression, FrameType, Literal, OffsetWindowFunction, RangeFrame, RowFrame, SpecifiedWindowFrame, UnboundedFollowing, UnboundedPreceding, UnspecifiedFrame, WindowExpression, WindowSpecDefinition}
+import org.apache.spark.sql.catalyst.expressions.aggregate.AggregateExpression
+import org.apache.spark.sql.types.{IntegerType, LongType}
+
+import io.substrait.`type`.Type
+import io.substrait.expression.{Expression => SExpression, ExpressionCreator, FunctionArg, WindowBound}
+import io.substrait.expression.Expression.WindowBoundsType
+import io.substrait.expression.WindowBound.{CURRENT_ROW, Following, Preceding, UNBOUNDED}
+import io.substrait.extension.SimpleExtension
+import io.substrait.relation.ConsistentPartitionWindow.WindowRelFunctionInvocation
+
+import scala.collection.JavaConverters
+
+abstract class ToWindowFunction(functions: Seq[SimpleExtension.WindowFunctionVariant])
+  extends FunctionConverter[SimpleExtension.WindowFunctionVariant, WindowRelFunctionInvocation](
+    functions) {
+
+  override def generateBinding(
+      sparkExp: Expression,
+      function: SimpleExtension.WindowFunctionVariant,
+      arguments: Seq[FunctionArg],
+      outputType: Type): WindowRelFunctionInvocation = {
+
+    val (frameType, lower, upper) = sparkExp match {
+      case WindowExpression(_: OffsetWindowFunction, _) =>
+        (WindowBoundsType.ROWS, UNBOUNDED, CURRENT_ROW)
+      case WindowExpression(
+            _,
+            WindowSpecDefinition(_, _, SpecifiedWindowFrame(frameType, lower, upper))) =>
+        (fromSpark(frameType), fromSparkPreceding(lower), fromSparkFollowing(upper))
+      case WindowExpression(_, WindowSpecDefinition(_, _, UnspecifiedFrame)) =>
+        (WindowBoundsType.ROWS, UNBOUNDED, CURRENT_ROW)
+      case _ => throw new UnsupportedOperationException(s"Unsupported window expression: $sparkExp")
+    }
+
+    ExpressionCreator.windowRelFunction(
+      function,
+      outputType,
+      SExpression.AggregationPhase.INITIAL_TO_RESULT, // use defaults...
+      SExpression.AggregationInvocation.ALL, // Spark doesn't define these
+      frameType,
+      lower,
+      upper,
+      JavaConverters.asJavaIterable(arguments)
+    )
+  }
+
+  def convert(
+      expression: WindowExpression,
+      operands: Seq[SExpression]): Option[WindowRelFunctionInvocation] = {
+    val cls = expression.windowFunction match {
+      case agg: AggregateExpression => agg.aggregateFunction.getClass
+      case other => other.getClass
+    }
+
+    Option(signatures.get(cls))
+      .filter(m => m.allowedArgCount(2))
```

Comment:
```java
m.allowedArgCount(2)
```
Is the intent to only filter out functions that take 2 arguments?

That line is also a no-op based on the implementation. Do we need this code?
```java
  def allowedArgCount(count: Int): Boolean = true
```

Reply:
You're right, I don't think it's needed either.  I'm guilty of cutting and pasting code that was originally part of the gluten project.  I'm sure it's there for historical reasons, but I'll remove it.